homeassistant/components/discovergy/__init__.py

```diff
@@ -0,0 +1,84 @@
+"""The Discovergy integration."""
+from __future__ import annotations
+
+from dataclasses import dataclass, field
+import logging
+
+import pydiscovergy
+from pydiscovergy.authentication import BasicAuth
+import pydiscovergy.error as discovergyError
+from pydiscovergy.models import Meter
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_EMAIL, CONF_PASSWORD, Platform
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryAuthFailed, ConfigEntryNotReady
+from homeassistant.helpers.httpx_client import get_async_client
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
+
+from .const import APP_NAME, DOMAIN
+
+PLATFORMS = [Platform.SENSOR]
+
+_LOGGER = logging.getLogger(__name__)
+
+
+@dataclass
+class DiscovergyData:
+    """Discovergy data class to share meters and api client."""
+
+    api_client: pydiscovergy.Discovergy = field(default_factory=lambda: None)
+    meters: list[Meter] = field(default_factory=lambda: [])
+    coordinators: dict[str, DataUpdateCoordinator] = field(default_factory=lambda: {})
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up Discovergy from a config entry."""
+    hass.data.setdefault(DOMAIN, {})
+
+    # init discovergy data class
+    discovergy_data = DiscovergyData(
+        api_client=pydiscovergy.Discovergy(
+            email=entry.data[CONF_EMAIL],
+            password=entry.data[CONF_PASSWORD],
+            app_name=APP_NAME,
+            httpx_client=get_async_client(hass),
+            authentication=BasicAuth(),
+        ),
+        meters=[],
+        coordinators={},
+    )
+
+    try:
+        # try to get meters from api to check if access token is still valid and later use
+        # if no exception is raised everything is fine to go
+        discovergy_data.meters = await discovergy_data.api_client.get_meters()
+    except discovergyError.InvalidLogin as err:
+        _LOGGER.debug("Invalid email or password: %s", err)
```

Comment:
We don't need to both log and raise the exception. The exception will be logged. We can remove this log.